pkg/cmd/cli/cmd/newapp.go

```diff
@@ -243,16 +243,26 @@ func (o *NewAppOptions) Run() error {
 		return err
 	}
 
-	// if the user has set the "app" label explicitly on their objects in the template,
-	// we should not return a failure when we can't set it ourselves.
-	ignoreLabelFailure := false
-	if len(config.Labels) == 0 && len(result.Name) > 0 {
-		config.Labels = map[string]string{"app": result.Name}
-		ignoreLabelFailure = true
+	// set labels explicitly supplied by the user on the command line
+	if err := setLabels(config.Labels, result, false); err != nil {
+		return err
 	}
 
-	if err := setLabels(config.Labels, result, ignoreLabelFailure); err != nil {
-		return err
+	// don't set the implicit "app" label if the user supplied an explicit list
+	// of labels to apply.
+	if len(config.Labels) == 0 && len(result.Name) > 0 {
+		// only set the computed implicit "app" label on objects if no object we've produced
+		// already has the "app" label.
+		appLabel := map[string]string{"app": result.Name}
+		hasAppLabel, err := hasLabel(appLabel, result)
+		if err != nil {
+			return err
+		}
+		if !hasAppLabel {
+			if err := setLabels(appLabel, result, true); err != nil {
```

Comment:
because short of parsing the text of the error message(ew) i can't currently distinguish why the addlabel call failed to idenitify "it failed because it existed"(what we care about) vs "it failed because something went horribly wrong"(which we'd need to catch and toss up the chain).

if only there were some sort of exception handling mechanism in which code identified the type of error it had encountered as part of throwing an error...